text/cluster-size-limits.md

```diff
@@ -0,0 +1,74 @@
+Cluster size limits
+==================
+
+## Summary
+
+One of the resource consumption limits for free-tier users is a cluster size limit.
+
+To enforce it, we need to calculate the timeline size and check if the limit is reached before relation create/extend operations.
+If the limit is reached, the query must fail some meaningful error/warning.
+We may want to exempt some operations from the quota to allow users to free space to fit back into the limit.
+
+Stateless compute node where the validation is performed is separate from the storage where the usage can be calculated, 
+so we need to exchange cluster size information between those components.
+
+## Motivation
+
+Enforce limit on the maximum size of a PostgreSQL instance to limit free tier users (and other tiers in the future).
+First of all this is needed to control the cost of our free tier production.
+Another reason to limit resourses is risk management — we haven’t (fully) tested and optimized zenith for really big clusters,
+so we don’t want to give users access to the functionality that we don’t think is ready.
+
+## Components
+
+pageserver - calculate size consumed by a timeline, add it into feedback message.
+safekeeper - pass feedback message from pageserver to compute.
+compute - receive feedback message, enforce size limit based on GUC `zenith.max_cluster_size`.
+console - set and update `zenith.max_cluster_size setting`
+
+## Proposed implementation
+
+First of all, it's necessary to define timeline size. Options are:
+
+1. Count all data, including SLRUs. (not including WAL)
+Here we think of it as of a physical disk underneath postgres cluster.
+This is how `LOGICAL_TIMELINE_SIZE` metric is implemented in pageserver right now and the draft PR uses it.
+
+2. Count only relation data. As in pg_database_size().
+
+This approach is more user-friendly, because it is the data that is really affected by user.
+To implement it, we will need to refactor timeline_size counter, or rather add another counter. 
```

Comment:
Currently LOGICAL TIMELINE SIZE doesn’t count everything, I remember that it doesn’t count TwoPhase files and as you mentioned wal is not counted too. So I think it can be just changed in place. Some time we will want to have a counter for data including wal. So I think two counters is enough